bagua/torch_api/data_parallel/distributed.py

```diff
@@ -0,0 +1,316 @@
+from torch.nn.modules import Module
+from torch._C._distributed_c10d import ProcessGroup as TorchProcessGroup
+from torch.nn.parallel import DistributedDataParallel as TorchDistributedDataParallel
+from typing import Callable, List, Optional, Union
+import warnings
+
+import bagua
+from bagua.torch_api.algorithms.gradient_allreduce import GradientAllReduceAlgorithm
+from contextlib import contextmanager
+from bagua.torch_api.communication import (
+    _get_default_group,
+    BaguaProcessGroup,
+)
+from .bagua_distributed import BaguaDistributedDataParallel
+
+
+class DistributedDataParallel_V1_9_0_Interface(Module):
+    r"""
+    PyTorch v1.9.0 DistributedDataParallel interface.
+    """
+
+    def __init__(self) -> None:
+        super(DistributedDataParallel_V1_9_0_Interface, self).__init__()
+
+    @contextmanager
+    def no_sync(self):
+        raise NotImplementedError
+
+    def forward(self, *inputs, **kwargs):
+        raise NotImplementedError
+
+    def scatter(self, inputs, kwargs, device_ids):
+        raise NotImplementedError
+
+    def to_kwargs(self, inputs, kwargs, device_id):
+        raise NotImplementedError
+
+    def gather(self, outputs, output_device):
+        raise NotImplementedError
+
+    def train(self, mode=True):
+        raise NotImplementedError
+
+    @contextmanager
+    def join(
+        self,
+        divide_by_initial_world_size=True,
+        enable=True,
+        throw_on_early_termination=False,
+    ):
+        raise NotImplementedError
+
+    def register_comm_hook(self, state: object, hook: Callable):
+        raise NotImplementedError
+
+    def will_sync_module_buffers(self):
+        raise NotImplementedError
+
+
+def to_bagua_process_group(process_group: Union[TorchProcessGroup, BaguaProcessGroup, None] = None):
+    if process_group is None:
+        return _get_default_group()
+    elif type(process_group) in [TorchProcessGroup, torch._C._distributed_c10d.ProcessGroupNCCL]:
+        return process_group.bagua_patch().bagua_pg  # pytype: disable=attribute-error
+    elif type(process_group) is BaguaProcessGroup:
+        return process_group
+    else:
+        raise Exception("unexpect input {}".format(type(process_group)))
+
+
+class DistributedDataParallel_V1_9_0(DistributedDataParallel_V1_9_0_Interface):
+    r"""
+    PyTorch v1.9.0 DistributedDataParallel interface using bagua backend.
+
+    :ivar bagua_optimizers: The optimizers constructor passed in.
+    :vartype bagua_optimizers: List[torch.optim.Optimizer]
+
+    :ivar bagua_algorithm: The algorithm implementation used by the module, reified by the algorithm constructor passed in.
+    :vartype bagua_algorithm: bagua.torch_api.algorithms.AlgorithmImpl
+
+    :ivar bagua_module_name: The module's name. Bagua uses the module name to distinguish different modules.
+    :vartype bagua_module_name: str
+
+    :ivar bagua_buckets: All Bagua buckets in a list.
+    :vartype bagua_buckets: List[bagua.torch_api.bucket.BaguaBucket]
```

Comment:
PyTorch v1.9.0 DistributedDataParallel (also add link here) compatible interface.

bagua_optimizers: Optimizer(s) used by the  module. It can contain one or more PyTorch optimizers. (The optimizers passed in are not constructorrs. We also  need to keep docs consistent with https://github.com/BaguaSys/bagua/blob/master/bagua/torch_api/distributed.py#L36)

bagua_algorithm, ...: same comment as bagua_optimizers